text/3759-cargo-supported-targets.md

```diff
@@ -0,0 +1,517 @@
+```
+```toml
+[package]
+name = "bar"
+
+[target.'cfg(target_os = "macos")'.dependencies]
+baz = "0.1.0"
+```
+Currently, `baz` is included in the dependency tree of `foo`, even though `foo` is never built for `macos`.
+With the addition of `supported-targets`, `baz` can be purged from the dependency tree of `foo`, since
+`target_os = "macos"` is mutually exclusive with `target_os = "linux"`.
+
+Formally, dependencies (and transitive dependencies) under `[target.**.dependencies]` tables are
+eliminated from the dependency tree of a package if the `supported-targets` of the package is mutually exclusive
+with the target preconditions of the dependency.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The `cfg` syntax is very expressive, but also very complex. As outlined above,
+    detecting subset and mutual exclusivity relations is not trivial.
+- Comparing `supported-targets` lists is an `O(n * m)` operation, with `n` and `m` being the number of elements
+    in the lists. (I doubt this will be a problem, as `n` and `m` are expected to be small).
+- Performance: this is a lot of extra calls to rustc.
+    Hopefully these are all compatible with the rustc cache, so they won't make things too bad.
+- This feature must be learned by those wanting to use dependencies with `supported-targets` specified.
+- This is the first step towards a target aware `cargo`, which may increase `cargo`'s complexity, and bring
+    more feature requests along these lines.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Format
+
+The list of strings format was chosen because of its simplicity and expressiveness. Other formats can also be considered:
+
+Using the `[target]` table, for example:
+```toml
+[target.'cfg(target_os = "linux")']
+supported = true
+```
+If the list of supported targets is long (should it ever be?), then the `Cargo.toml` file becomes very verbose
+as well.
+
+A `[suppported]` table, with `arch = ["<arch>", ...]`, `os = ["<os>", ...]`, `target = ["<target>", ...]`, etc.
+This is more verbose, complex to implement, learn, and remember. It is also not obvious how `not` and `all`
+could be represented in this format. For example:
+```toml
+[supported]
+os = ["linux", "macos"]
+arch = ["x86_64"]
+```
+
+## Naming
+
+Some other names for this field can be considered:
+
+- `required-targets`. Pro: it matches with the naming of `required-features`. Con: `required-features` is a list of features
+    that must _all_ be enabled (conjunction), whereas `supported-targets` is a list of targets
+    where _any_ is allowed (disjunction).
+- `targets`. As in "this package _targets_ ...". Pro: Concise. Con: Ambiguous, and could be confused with the `target` table.
+
+## Package scope vs. cargo-target scope
+
+The `supported-targets` field is placed at the package level, and not at the cargo-target
+level (i.e., under, `[lib]`, `[[bin]]`, etc.). A rationale is given for why the cargo-target level is not used.
```

Comment:
```suggestion
The `supported-targets` field is placed at the package level, and not at the cargo-target
level (i.e., under, `[lib]`, `[[bin]]`, etc.)
```
This is a bit redundant as the intention of having a section like this is to provide the rationale.  Having it here makes it read awkwardly to me, including that it implies its going to say something different from the norm but it doesn't.